File: completable-reactor-runtime/src/main/java/ru/fix/completable/reactor/runtime/gl/GlTransitionBuilderImpl.java
package ru.fix.completable.reactor.runtime.gl;

final class GlTransitionBuilderImpl implements GlTransitionBuilder {

    final Vertex vertex;
    final GlTransition transition;

    public GlTransitionBuilderImpl(Vertex vertex, GlTransition transition) {
        this.vertex = vertex;
        this.transition = transition;
    }

    @Override
    public Vertex complete() {
        transition.setComplete(true);
        return vertex;
    }

    @Override
    public Vertex handleBy(Vertex vertex) {
        transition.setHandleBy(vertex);
        return vertex;
    }

    @Override
    public Vertex mergeBy(Vertex vertex) {
        transition.setMergeBy(vertex);
        return vertex;
    }
}

File: completable-reactor-runtime/src/main/kotlin/ru/fix/completable/reactor/runtime/gl/GlTransitionBuilderImpl.kt
package ru.fix.completable.reactor.runtime.gl

import ru.fix.completable.reactor.runtime.internal.gl.ConfigContext

class GlTransitionBuilderImpl(
        val vertex: Vertex,
        val transition: GlTransition) : GlTransitionBuilder {

    override fun complete(): Vertex {
        transition.isComplete = true
        return vertex
    }

    override fun handleBy(vertex: Vertex): Vertex {
        if (vertex.name == null) {
            vertex.name = ConfigContext.get().resolveVertexName(vertex)
        }
        transition.handleBy = vertex
        return vertex
    }

    override fun mergeBy(vertex: Vertex): Vertex {
        if (vertex.name == null) {
            vertex.name = ConfigContext.get().resolveVertexName(vertex)
        }

        if (vertex.merger == null) {
            throw IllegalArgumentException("""
                MergeBy transition is targeting vertex ${vertex.name} that does not have merger.
                MergeBy can only transition to vertexes with mergers.
                You can build such vertexes in two manners:
                    handler(...).withMerger(...)
                    subgraph(...).withMerger(...)
                Other vertices that built without mergers can not be targeted by mergeBy transition.
            """.trimIndent())
        }

        transition.mergeBy = vertex
        return vertex
    }
}
